autogpt_platform/backend/backend/blocks/google/gmail.py

```diff
@@ -393,37 +437,238 @@ def __init__(self):
         )
 
     async def run(
-        self, input_data: Input, *, credentials: GoogleCredentials, **kwargs
+        self,
+        input_data: Input,
+        *,
+        credentials: GoogleCredentials,
+        graph_exec_id: str,
+        user_id: str,
+        **kwargs,
     ) -> BlockOutput:
         service = GmailReadBlock._build_service(credentials, **kwargs)
         result = await asyncio.to_thread(
             self._send_email,
             service,
-            input_data.to,
-            input_data.subject,
-            input_data.body,
+            input_data,
+            graph_exec_id,
+            user_id,
         )
         yield "result", result
 
-    def _send_email(self, service, to: str, subject: str, body: str) -> dict:
-        if not to or not subject or not body:
-            raise ValueError("To, subject, and body are required for sending an email")
-        message = self._create_message(to, subject, body)
+    def _send_email(
+        self, service, input_data: Input, graph_exec_id: str, user_id: str
+    ) -> dict:
+        if (
+            not input_data.to
+            or len(input_data.to) == 0
+            or not input_data.subject
+            or not input_data.body
+        ):
+            raise ValueError(
+                "At least one recipient, subject, and body are required for sending an email"
+            )
+        message = self._create_message(input_data, graph_exec_id, user_id)
         sent_message = (
             service.users().messages().send(userId="me", body=message).execute()
         )
         return {"id": sent_message["id"], "status": "sent"}
 
-    def _create_message(self, to: str, subject: str, body: str) -> dict:
+    def _create_message(
+        self, input_data: Input, graph_exec_id: str, user_id: str
+    ) -> dict:
+        from email import encoders
+        from email.mime.base import MIMEBase
+        from email.mime.multipart import MIMEMultipart
         from email.mime.text import MIMEText
 
-        message = MIMEText(body)
-        message["to"] = to
-        message["subject"] = subject
+        message = MIMEMultipart()
+        # Handle both string and list for 'to' field, including JSON strings
+        message["to"] = parse_email_recipients(input_data.to)
+        message["subject"] = input_data.subject
+
+        if input_data.cc:
+            message["cc"] = ", ".join(input_data.cc)
+        if input_data.bcc:
+            message["bcc"] = ", ".join(input_data.bcc)
+
+        message.attach(MIMEText(input_data.body))
+
+        # Handle attachments if any
+        if input_data.attachments:
+            import asyncio
+
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
```

Comment:
Will this cause an issue on multiple calls?  Why don't you make this function async? blocks are async functions